gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageGrpcWriteChannel.java

```diff
@@ -177,7 +180,7 @@ private class UploadOperation implements Callable<WriteObjectResponse> {
     private InsertChunkResponseObserver responseObserver;
     // Holds list of most recent number of NUMBER_OF_REQUESTS_TO_RETAIN requests, so upload can
     // be rewound and re-sent upon transient errors.
-    private final TreeMap<Long, ByteString> dataChunkMap = new TreeMap<>();
+    private final TreeMap<Long, WriteObjectRequest> dataChunkMap = new TreeMap<>();
```

Comment:
Let's avoid making multiple changes in a single PR. Can we split this change

1. Dropping buffered data
2. Caching requests instead of ByteStrings -- I suspect this change can be avoided now if we reconsider the design itself skipping the extra memory copy
3. Retry handling with TRANSIENT_ERRORS

Reply:
I think all of these are related. The PR anyway logs the commit history, so although the branch-2.x.x will have one commit, the history will be preserved. Splitting to 3 PR will require a lot more work.